src/main.cpp

```diff
@@ -2265,209 +2163,287 @@ void MainFrame::startRxStream()
-                delete m_rxOutPa;
-                if(two_tx)
-                    delete m_txOutPa;
-                destroy_fifos();
-                destroy_src();
-                deleteEQFilters(g_rxUserdata);
-                delete g_rxUserdata;
-                m_RxRunning = false;
-                Pa_Terminate();
-                return;
-            }
-
-            m_rxErr = m_rxOutPa->streamStart();
-            if(m_rxErr != paNoError) {
-                wxMessageBox(wxT("Sound Card 1 Second Stream Start Error."), wxT("Error"), wxOK);
-                m_rxInPa->stop();
-                m_rxInPa->streamClose();
-                delete m_rxInPa;
-                delete m_rxOutPa;
-                if(two_tx)
-                    delete m_txOutPa;
-                destroy_fifos();
-                destroy_src();
-                deleteEQFilters(g_rxUserdata);
-                delete g_rxUserdata;
-                m_RxRunning = false;
-                Pa_Terminate();
-                return;
+        // Set sound card callbacks
+        auto errorCallback = [&](IAudioDevice&, std::string error, void*)
+        {
+            CallAfter([&]() {
+                wxMessageBox(wxString::Format("Error encountered while processing audio: %s", error), wxT("Error"), wxOK);
+            });
+        };
+
+        rxInSoundDevice->setOnAudioData([](IAudioDevice& dev, void* data, size_t size, void* state) {
+            paCallBackData* cbData = static_cast<paCallBackData*>(state);
+            short* audioData = static_cast<short*>(data);
+            short  indata[size];
+            for (int i = 0; i < size; i++, audioData += dev.getNumChannels())
+            {
+                indata[i] = audioData[0];
             }
-        }
-
-        if (g_verbose) fprintf(stderr, "started stream 1\n");
-
-        // Start sound card 2 ----------------------------------------------------------
-
-        if (g_nSoundCards == 2) {
-
-            // question: can we use same callback data
-            // (g_rxUserdata)or both sound card callbacks?  Is there a
-            // chance of them both being called at the same time?  We
-            // could need a mutex ...
-
-            m_txInPa->setUserData(g_rxUserdata);
-            m_txErr = m_txInPa->setCallback(txCallback);
-            m_txErr = m_txInPa->streamOpen();
-
-            if(m_txErr != paNoError) {
-                if (g_verbose) fprintf(stderr, "Err: %d\n", m_txErr);
-                wxMessageBox(wxT("Sound Card 2 Open/Setup error."), wxT("Error"), wxOK);
-                m_rxInPa->stop();
-                m_rxInPa->streamClose();
-                delete m_rxInPa;
-                if(two_rx) {
-                    m_rxOutPa->stop();
-                    m_rxOutPa->streamClose();
-                    delete m_rxOutPa;
+            
+            {
+                std::unique_lock<std::mutex> lk(cbData->infifo1Mutex);
+                if (codec2_fifo_write(cbData->infifo1, indata, size)) 
```

Comment:
As long as there is only one writer and one reader process, the codec2 FIFOs are [thread safe by design](https://github.com/drowe67/codec2/blob/master/unittest/tfifo.c) and mutexes are not required

Reply:
Ah, okay. I had run into an issue last night during my own testing where it seemed to get stuck repeatedly playing the last ~50-100ms of audio. I think forcing high latency on PortAudio resolved it and not the mutexes, though, so I'll remove.